### PR TITLE
Map HUD

### PR DESCRIPTION
**Pull request by Warriorrrr**
#### Description: 

Adds a new map HUD using code from the pre-adventure ascii map.

Example image:
image

____
#### New Nodes/Commands/ConfigOptions: 

____
#### Relevant Towny Issue ticket:

____
- [x] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

File: src/com/palmergames/bukkit/towny/huds/HUDManager.java [hunk 1]
```diff
@@ -93,8 +106,15 @@ public void onPlayerMovePlotsEvent(PlayerChangePlotEvent event) throws NotRegist
...
-			
+			}	
+		} else if (mapUsers.contains(p)) {
+			if (!isMapHudActive(p))
+				mapUsers.remove(p);
+			else	
+				if (event.getTo().getTownyWorld().isUsingTowny())
+					MapHUD.updateMap(p, event.getTo());
+				else
+					toggleAllOff(p);
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/huds/HUDManager.java:117]: else-if else?

File: src/com/palmergames/bukkit/towny/huds/MapHUD.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	private static int lineWidth = 20, lineHeight = 10;
+	
+	public static void toggleOn(Player player) {
+		Scoreboard board = Bukkit.getScoreboardManager().getNewScoreboard();
+		Objective objective = board.registerNewObjective("MAP_HUD_OBJ", "", "");
+		objective.setDisplaySlot(DisplaySlot.SIDEBAR);
+		
+		int score = lineHeight + 2;
+		for (int i = 0; i < lineHeight; i++) {
+			board.registerNewTeam("mapTeam" + i).addEntry(ChatColor.values()[i].toString());
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/huds/MapHUD.java:31]: Store `ChatColor.values()` in a variable. `values()` will allocate a new array every call.

File: src/com/palmergames/bukkit/towny/huds/MapHUD.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			objective.getScore(ChatColor.values()[i].toString()).setScore(score);
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/huds/MapHUD.java:32]: See above

File: src/com/palmergames/bukkit/towny/huds/MapHUD.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		
+		String[][] map = new String[lineWidth][lineHeight];
+		int x, y = 0;
+		for (int tby = wc.getX() + (lineWidth - halfLineWidth - 1); tby >= wc.getX() - halfLineWidth; tby--) {
+			x = 0;
+			for (int tbx = wc.getZ() - halfLineHeight; tbx <= wc.getZ() + (lineHeight - halfLineHeight - 1); tbx++) {
+				try {
+					TownBlock townblock = world.getTownBlock(tby+1, tbx);
+					if (!townblock.hasTown())
+						throw new TownyException();
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/huds/MapHUD.java:91]: I feel like throwing an exception is a redundant overhead when you can just perform the logic inside the if-statement.

File: src/com/palmergames/bukkit/towny/huds/MapHUD.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+									// towns
+									map[y][x] = Colors.Red;
+								else
+									map[y][x] = Colors.White;
+							} else
+								map[y][x] = Colors.White;
+						} else
+							map[y][x] = Colors.White;
+					} else
+						map[y][x] = Colors.White;
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/huds/MapHUD.java:123]: Is the really no cleaner way to do this...

**Review by silverwolfg11 — COMMENTED** [review]
Super excited for this feature! I hate when I'm moving around and my chat gets spammed with the ASCII map, so much appreciated.   

Just some really nit-picky things, but other than those looks good to me! 😄

**Review by LlmDl — APPROVED** [review]

**Review by silverwolfg11 — APPROVED** [review]

**Pull request merged**